### DISCUSSION TITLE
How to profiling CPU and memory usage

### DISCUSSION THREAD

**Discussion by pusponegorow** [Kong/kong#13871] [Help] [answered]
Hi Kong community,

I'm looking to understand the best approaches for profiling Kong Gateway's CPU and memory usage. Specifically, I'd like to:

1. Monitor and analyze CPU usage patterns
2. Track memory consumption and potential leaks
3. Identify performance bottlenecks

What are the recommended tools and methods for profiling Kong? I'm particularly interested in:
- Built-in Kong profiling capabilities
- Third-party profiling tools that work well with Kong
- Best practices for production environment profiling
- Key metrics to monitor

Any examples or documentation references would be greatly appreciated.

Thanks in advance!

**Comment by xianghai2** [comment, marked as answer]
Kong Enterprise offers a feature for this:
https://docs.konghq.com/gateway/latest/reference/cli/

The CLI "Kong Debug" command could be used to profiling cpu/memory and gc-snapshot etc.

**Reply by pusponegorow** [reply]
yes i'm aware for kong enterprise.

but is there workaround for profiling kong open source?

**Reply by guanlan** [reply]
You can try the stapxx tool from OpenResty at your own risk: https://github.com/openresty/stapxx/tree/master?tab=readme-ov-file#lj-lua-stacks
